internal/controller/controllers/clusterdeployments_controller.go

```diff
@@ -1538,6 +1550,51 @@ func (r *ClusterDeploymentsReconciler) getNumOfClusterAgents(c *common.Cluster)
 	return r.Installer.GetKnownHostApprovedCounts(*c.ID)
 }
 
+func findAgentsByClusterDeployment(k8sclient client.Client, ctx context.Context, log logrus.FieldLogger, cdName, cdNamespace string) ([]aiv1beta1.Agent, error) {
+	agentList := aiv1beta1.AgentList{}
+	agents := []aiv1beta1.Agent{}
+	// TODO(chocobomb) Check if ctx does not limit the List() call only to the CD's namespace
+	//                 because Agents for a CD may live in another NS, so that would be an issue.
+	err := k8sclient.List(ctx, &agentList, client.MatchingLabels{aiv1beta1.Group + "/clusterdeployment-namespace": cdNamespace})
+
+	if err != nil {
+		return agents, err
+	}
+
+	for _, agent := range agentList.Items {
+		if agent.Spec.ClusterDeploymentName == nil {
+			continue
+		}
+		if agent.Spec.ClusterDeploymentName.Name != cdName {
+			continue
+		} else {
+			agents = append(agents, agent)
+		}
+	}
+
+	return agents, nil
+}
+
+func (r *ClusterDeploymentsReconciler) getNumOfSyncedClusterAgents(ctx context.Context, log logrus.FieldLogger, aci *hiveext.AgentClusterInstall) (int, error) {
+	cdName := aci.Spec.ClusterDeploymentRef.Name
+	cdNamespace := aci.Namespace
+	agents, err := findAgentsByClusterDeployment(r.Client, ctx, log, cdName, cdNamespace)
```

Comment:
Does it make sense to move this call up to the main reconcile method?  My concern is that this method is fairly expensive and in the future we may have another use for it, so moving it to the main reconcile method might increase the chances that whoever implements that won't call the method again.